### ISSUE TITLE
Custom HTML block adding extra characters

### ISSUE THREAD

**Issue by bradhogan** [WordPress/gutenberg#49166]
### Description

Tough to nail down exactly when this occurs, but it seems like it happens any time my Front Page template is saved. I've removed ALL whitespace even and random `</p>` and `<br /> `tags are added again and again.  Note the odd closing </p> in this screenshot and random `<p>` and `<br />`.  Again, I've removed those, then removed the whitespace, saved and it works for the immediate moment, then something happens when I come back and make updates in the Site Editor and the odd characters reappear.

### Step-by-step reproduction instructions

1. Add custom HTML block to a template 
```
<form class="has-text-align-center subscribe-form" action="" data-remote="true"> <input type="text" name="first_name" placeholder="First name..."><input type="email" name="email" placeholder="Your@email">
<div class="ck_control_group ck_captcha2_h_field_group ck-captcha2-h" style="position: absolute !important;left: -999em !important;"><label class="ck_label" for="ck_captcha2_h">We use this field to detect spam bots. If you fill this in, you will be marked as a spammer.</label><input type="text" name="captcha2_h" class="ck-captcha2-h" id="ck_captcha2_h"></div>
<button class="wp-element-button">Join the waitlist</button>
</form>
```

2. Save, then logout and log back in, make changes in the Site Editor, save, then go back to edit the template, make changes, exit out, go back and see if the custom html block now has extra tags in there

### Screenshots, screen recording, code snippet

<img width="665" alt="Screen Shot 2023-03-17 at 10 40 10 AM" src="https://user-images.githubusercontent.com/1781226/225967452-c39e25bf-8fba-478e-8936-bbff0a067191.png">

### Environment info

- wp 6.1.1
- gutenberg 15.3.1
- chrome
- mac

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by Thelmachido** [comment]
<img width="717" alt="Screenshot 2023-03-20 at 14 47 44" src="https://user-images.githubusercontent.com/41231764/226277981-18b8dfbf-167e-430f-a49d-e6a26acffc8a.png">

 In the screenshot below you can see how the waiting list form's button has moved to the bottom, and the screen size for the editor wasn't adjusted I continuously closed and opened the template and a <p> tag was added to the custom HTML. 

<img width="765" alt="Screenshot 2023-03-20 at 14 56 17" src="https://user-images.githubusercontent.com/41231764/226279325-172b8e8a-4c29-40ed-83e8-28e21ab0a340.png">

WP 6.1.1 
Gutenberg 15.3.1
TT3 theme

**Comment by Marielle83** [comment]
Edit: updating my comment with steps and couple of images.

WordPress version: 6.2 RC4
No plugins installed, using Twenty Twenty-Three Theme, but happens with any block theme.

Open the Site Editor and edit a template, for example Page template. Insert Custom HTML block somewhere and put in some HTML. Save and refresh the page (or save and exit Editor then come back to edit the template) The HTML now has added `<p>` and `<br>` tags added.

I'll add some screenshots as an example showing it also happens if you add scripts in the Custom HTML block. Probably wouldn't need the Custom HTML block for this, but it's just an example.

[Intact HTML](https://i.imgur.com/oj0FVy1.png)
[Destroyed HTML and script](https://i.imgur.com/mdI8SQK.png)

Expected result should be the HTML and script being intact, without added tags.

The issue isn't currently in 6.1.1, but if using the Gutenberg development plugin (the one that says "This is the development plugin for the block editor, site editor, and other future WordPress core functionality.") in 6.1.1 the issue is there though and I can reproduce it using the steps above.

**Comment by connah99** [comment]
Thought I'd chime in since this is affecting me too. As @bradhogan said, the HTML tags are only added once the Editor is re-entered after saving. Strangely, once I re-enter the editor, I don't need to save and exit for the HTML to be altered. Not all HTML Blocks are affected, and the changes aren't just limited to adding tags, but also messing up the entire hierarchy and layout. I have an example of the one affected here, where the top is the broken HTML and the bottom is the original: 

<img width="1222" alt="Screenshot 2023-03-25 at 21 28 00" src="https://user-images.githubusercontent.com/91231626/227743274-d8857266-f36b-48e5-bcbd-796b7fd8b352.png">

**Comment by alemanriquezr** [comment]
Happening to me too, after saving and reloading too.
WP version 6.2, twenty twenty three. The < p > and < br / > tags are added in CSS input in custom html block < style > tags. example:
< style >
< p >.class{ < br />
	rule;< br />
	rule;< br />
} < / p >
< / style >

**Comment by TobiasBg** [comment]
I also have been seeing this for a few weeks now. My HTML code in the block is perfectly valid.
Indeed, it looks like something like `wpautop` affecting this.

**Comment by perrytrademark** [comment]
I can confirm that. Thats a legit bug that needs to be fixed. As an alternative i am using another plugin that provides a custom HTML block
![strange issue](https://user-images.githubusercontent.com/22387795/235079557-449b1cfe-c4fc-42b0-a964-e00b0889eb41.jpg)

**Comment by Rale379** [comment]
Same issue. Each time the template is uploaded / opened, the tags are added.
I use the block for adding CSS that is specific to that template.

**Comment by justintadlock** [comment]
I ran some rudimentary tests on this, but this seems to only be an issue when editing a template from the site editor (**Appearance > Editor > Templates**). When editing the Custom HTML block from the post editor or the template editor (via the post editing screen), the issue didn't happen.

![template editor](https://user-images.githubusercontent.com/1816309/235166004-2972ee2e-e065-478c-ac3b-d7664c3bfd04.png)

![site-editor](https://user-images.githubusercontent.com/1816309/235166012-eeaa828b-5cfe-4b37-8946-f0372285fc5d.png)

**Comment by Rale379** [comment]
@justintadlock 
So, you're saying it happens on templates, but not on posts and pages

**Comment by perrytrademark** [comment]
> @justintadlock So, you're saying it happens on templates, but not on posts and pages

no the bug happens to normal pages and posts also. Not only in the template editor. Also it's a 6.2 issue, in 6.1 it was fine.

**Comment by dalebridge** [comment]
> I can confirm that. Thats a legit bug that needs to be fixed. As an alternative i am using another plugin that provides a custom HTML block
> ![strange issue](https://user-images.githubusercontent.com/22387795/235079557-449b1cfe-c4fc-42b0-a964-e00b0889eb41.jpg)

I am having the same issue. Out of interest, which plugin are you using in the mean time?

**Comment by perrytrademark** [comment]
> > I can confirm that. Thats a legit bug that needs to be fixed. As an alternative i am using another plugin that provides a custom HTML block
> > ![strange issue](https://user-images.githubusercontent.com/22387795/235079557-449b1cfe-c4fc-42b0-a964-e00b0889eb41.jpg)
> 
> I am having the same issue. Out of interest, which plugin are you using in the mean time?

This one "https://wordpress.org/plugins/advance-custom-html/". It provides code highlight also which is very usefull.

**Comment by matthewbullweb** [comment]
Adding P and Br tags Bug is still present in custom html block in FSE template editor today.

**Comment by obache** [comment]
Put followings in the editor:
```html
<a class="button" download href="/dummy.pdf">

 <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 12">
  <g></g>
 </svg>

</a>
```
Result in:
```html
<a class="button" download="" href="/dummy.pdf"></a>

<a class="button" download="" href="/dummy.pdf"> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 12"></svg>
<g></g>
</a><a class="button" download="" href="/dummy.pdf"></a>
```

It's so terrible formatting.

**Comment by cdefy** [comment]
This happened for me after the 6.2.2 WP update (the one to fix the removal of short code handling in template!), it didn't do this on 6.2.0 or 6.2.1 :(

Any word on that? It's annoying as it breaks the display of some shortcode :(

**Comment by matthewbullweb** [comment]
I found the following filter code online. Didn't appear to do anything for me. Maybe others will have more luck,

`// Prevent WP from adding <p> tags on all post types
function disable_wp_auto_p( $content ) {
  remove_filter( 'the_content', 'wpautop' );
  remove_filter( 'the_excerpt', 'wpautop' );
  remove_filter('widget_custom_html', 'wpautop' , 10, 3 ); //new line
  return $content;
}

add_filter( 'the_content', 'disable_wp_auto_p', 0 );`

Reference:
https://templ.io/blog/prevent-wordpress-from-adding-p-tags-in-content/#:~:text=If%20you%20want%20to%20prevent,your%20post%20type%20of%20choice.

**Comment by perrytrademark** [comment]
Yes this is correct. Random p and br tags are now inserted when a shortcode is present. This is a mess, I reverted WordPress back to previous versions till this is fixed.-------- Original message --------From: cdefy ***@***.***>Date: Thu, Jun 1, 2023, 11:54To: WordPress/gutenberg ***@***.***>Cc: perrytrademark ***@***.***>, Comment ***@***.***>Subject: Re: [WordPress/gutenberg] Custom HTML block adding extra characters (Issue #49166)
This happened for me after the 6.2.2 WP update (the one to fix the removal of short code handling in template!), it didn't do this on 6.2.0 or 6.2.1 :(
Any word on that? It's annoying as it breaks the display of some shortcode :(

**Comment by cdefy** [comment]
@matthewbullweb Didn't work for me neither.

But found this other workaround and it works for me: https://core.trac.wordpress.org/ticket/58366#comment:11

In `wp-includes/block-template.php`, after the line `$content = str_replace( ']]>', ']]&gt;', $content );` I added:
```
$content = str_replace( '<p>', '', $content );
$content = str_replace( '</p>', '', $content );
```

Hope it can help others while waiting for WP to fix this. :(

**Comment by ghost** [comment]
Great to see this issue in the Todo for WP 6.3, as this was a serious regression in my opinion in WP 6.2.
Currently it is not possible to apply custom CSS directly to Templates via Theme.json. Because of this, the HTML Block was considered first in my community as the way to add custom CSS to templates. But now the HTML block is unfortunately broken inside templates.

**Comment by andrewserong** [comment]
I'm not sure if it covers all of the cases that folks have described so far in this issue, but I have a tentative fix open in #52716 if anyone would like to take it for a spin (addresses the issue within templates in the site editor).
